generator/internal/codegen_utils.cc

```diff
@@ -87,6 +88,44 @@ std::string ProtoNameToCppName(absl::string_view proto_name) {
   return "::" + absl::StrReplaceAll(proto_name, {{".", "::"}});
 }
 
+std::vector<std::string> BuildNamespaces(
+    std::map<std::string, std::string> const& vars, NamespaceType ns_type) {
+  // vars["product_path"] is guaranteed to be present and properly formatted.
+  std::string product_path = vars.find("product_path")->second;
+  std::vector<std::string> v = absl::StrSplit(product_path, '/');
+  auto name = v[v.size() - 2];
+  std::string inline_ns = absl::AsciiStrToUpper(name) + "_CLIENT_NS";
+  if (ns_type == NamespaceType::INTERNAL) {
+    name = absl::StrCat(name, "_internal");
+  }
+  return {"google", "cloud", name, inline_ns};
+}
+
+StatusOr<std::vector<std::pair<std::string, std::string>>>
+ProcessCommandLineArgs(std::string const& parameters) {
+  std::cerr << "paramters: " << parameters << std::endl;
+  std::vector<std::pair<std::string, std::string>> command_line_args;
+  google::protobuf::compiler::ParseGeneratorParameter(parameters,
+                                                      &command_line_args);
+
+  auto product_path =
+      std::find_if(command_line_args.begin(), command_line_args.end(),
+                   [](std::pair<std::string, std::string> const& p) {
+                     return p.first == "product_path";
+                   });
+  if (product_path == command_line_args.end() || product_path->second.empty()) {
+    return Status(StatusCode::kInvalidArgument,
+                  "--cpp_codegen_opt=product_path=<path> must be specified.");
+  }
```

Comment:
`auto& path = product_path->second;`
... then you can just use `path` in the lines below.